### PR TITLE
Feature/847/capi fleetcarrier

### PR DESCRIPTION
**Pull request by aussig**
This PR implements #847 - calling of the CAPI endpoint for fleetcarrier data. 

This call is only triggered on `CarrierBuy`, `CarrierStats` and `CarrierTradeOrder` journal events, and is on a separate cooldown (currently 15 mins) to the main CAPI call.

The new plugin callback function is `capi_fleetcarrier()`.

### CODE REVIEW

File: plug.py
Comment (Athanasius) [inline plug.py:404]: I don't think `is_beta` is actually necessary in this case.  The `.source_host` should denote it (because if there's ever a beta of the game it'll be `pts-companion.orerve.net`), and if somehow the beta detection has failed and the live host was used anyway, then the data will be for the live game, not beta.

File: EDMarketConnector.py
Comment (Athanasius) [inline EDMarketConnector.py:1568]: Re: `CargoTransfer` ... I'm conflicted on this because whilst it's a pro to trigger the query after cargo hold changes ... doing so on the first of possibly many would 'jump the gun' and only updated for the first... and that's assuming the CAPI data updates in a timely manner.

The latter also applies to `CarrierTradeOrder`.  I'm leaning towards requiring the player to relogin in the game (to get `CarrierStats`) as the simplest way to give clear player control for this, without complicating things with an extra UI button/config options.

File: PLUGINS.md
Comment (Athanasius) [inline PLUGINS.md:945]: Add a note here about how a single query can take a long time under some circumstances.  This means not only could a fleetcarrier update take a long time, *but it will also possibly delay all other CAPI queries*.

File: companion.py
Comment (Athanasius) [inline companion.py:959]: This needs to specify a custom `timeout=...`, because the default 10s isn't going to work for the slow queries.

The query is threaded, so won't hold up the EDMC UI, but can hold up other CAPI queries, so I'm inclined to try this with 60s for now.

Comment (Athanasius) [reply companion.py:959]: Long-term we might need to adjust `capi_query_worker()` to actually perform *these* (or future other 'not station' query set) queries in their own sub-thread.  But that can wait until we see how *this* shakes out in practice.

File: PLUGINS.md
Comment (Athanasius) [reply PLUGINS.md:945]: Also, now I realise we might have lost some "don't allow 'Update' if already in progress" check when I made CAPI threaded.  But don't worry about that for this PR.

**Review by Athanasius — CHANGES_REQUESTED** [review]
Mostly minor changes needed, but some further discussion about some things required.

File: EDMarketConnector.py
Comment (A-UNDERSCORE-D) [reply EDMarketConnector.py:1096]: additionally if you're not going to pass data to be modified, you should use `_` for `new_data` to make that visible

File: EDMarketConnector.py
Comment (A-UNDERSCORE-D) [inline EDMarketConnector.py:1568]:
```suggestion
            if entry['event'] in ('CarrierBuy', 'CarrierStats', 'CarrierTradeOrder'):  # 'CargoTransfer' too?
```

While its not likely to be changed / changable, good practice to use tuples here

**Review by A-UNDERSCORE-D — CHANGES_REQUESTED** [review]

File: EDMarketConnector.py
Comment (Athanasius) [reply EDMarketConnector.py:1568]: Re: using `_` ... to be fair, that wasn't caught when *I* recently added the CAPI killswitches.  It'll need changing in general for those.

File: plug.py
Comment (Athanasius) [reply plug.py:421]: Yes, they're both still things, and we ought to put that `callable(...)` check on all the plugin calling code in plug.py.

FYI @aussig the deep copy is so that one plugin mutating the data doesn't affect a subsequent one.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1096]: I'll implement both changes.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1109]: I originally excluded the CQC and Multicrew checks as I agree they are irrelevant, but thought a check for a valid commander might be a good idea. Will remove that too though.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1123]: I'll implement.

File: EDMarketConnector.py
Comment (Athanasius) [reply EDMarketConnector.py:1109]: It's the `monitor.mode` check I was referring to.

You need the `monitor.cmdr` check, else it's going to fail anyway when it can't determine which Cmdr's credentials to use for the query.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1568]: Ok, will remove that comment and also remove `CarrierTradeOrder` as a trigger.

File: plug.py
Comment (aussig) [reply plug.py:404]: Yeah, was following the pattern of the other plugin callbacks, but agree everything is in the `source_host`. Will change.

File: PLUGINS.md
Comment (aussig) [reply PLUGINS.md:945]: Will add some additional text.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1568]: Want me to do that @Athanasius ?

Comment (Athanasius) [reply EDMarketConnector.py:1568]: Yes, it's the more correct form.  In general if A_D requests something and I haven't objected, take it as canon.  They know more Python than I do ;) .

File: plug.py
Comment (aussig) [reply plug.py:421]: Will implement both.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1109]: Ah ok, that makes more sense.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1568]: Ok, tried this and it immediately throws an exception:

```
self.ship_label['text'] = _('Ship') + ':'  # LANG: 'Ship' label in main UI
                              ^
UnboundLocalError: cannot access local variable '_' where it is not associated with a value
```

This exception isn't thrown anywhere near any of the instances of 
```python
should_return, new_data = killswitch.check_killswitch('capi.request.fleetcarrier', {})
```
that I changed to:
```python
should_return, _= killswitch.check_killswitch('capi.request.fleetcarrier', {})
```

So, I suspect this is because `_()` is defined by i18n [as described here](https://stackoverflow.com/questions/5893163/what-is-the-purpose-of-the-single-underscore-variable-in-python/5893946#5893946)

That article suggests using double underscore `__` as an alternative for throwaway variables when used in conjunction with l18n.  @A-UNDERSCORE-D would that be acceptable?

Comment (Athanasius) [reply EDMarketConnector.py:1568]: Yup, if `__` works, then we'll have to use that (and I'm surprised where I changed it apparently hasn't blown up yet in this branch, but I've only just gotten to test-running it).

Comment (aussig) [reply EDMarketConnector.py:1568]: Ok, will stick with that.

**Comment by aussig** [comment]
Thanks all, I've implemented the changes from the first round of reviews.

**Comment by Athanasius** [comment]
I think everything in my Review has been address, however...

I just had a lovely edge-case thought:

1. Player with two accounts, each has an FC.
2. Login into Cmdr A, `/fleetcarrier` query issued, but it's a slow one.
3. Meantime, re-logs to Cmdr B.
4. Response from #2 arrives.

Now, the way it's currently architected `capi_fleetcarrier()` doesn't make any guarantees about the data being for "the current Cmdr", so technically this is OK.  But plugins would have to be aware of this and cater for it.

But, perhaps, we should record the `monitor.cmdr` at query time (I'd expand EDMCCAPIRequest to have this as a constructor arg and store it, to then be mirrored in the EDMCCAPIRespons), such that it's an extra property in CAPIData when passed to plugins.

**Comment by aussig** [comment]
> But, perhaps, we should record the `monitor.cmdr` at query time (I'd expand EDMCCAPIRequest to have this as a constructor arg and store it, to then be mirrored in the EDMCCAPIRespons), such that it's an extra property in CAPIData when passed to plugins.

Ok, I'll try to get this done today, though it may have to wait to after the holidays as it's about to get very busy here!

**Comment by Athanasius** [comment]
No rush, we're not planning on 5.8.0 any time soon.  And there's a bunch of 'small' issues in the milestone that I want to try and get cleared for it.

But, I'll probably end up merging my "get *everything* passing mypy checks" PR before you're done now, and then you'll definitely want to rebase, especially given you've applied changes in files irrelevant to your actual work (the plugins).

**Review by Athanasius — APPROVED** [review]

**Review by A-UNDERSCORE-D — APPROVED** [review]

**Comment by Athanasius** [comment]
> If I inject this into the `CAPIData` itself, should I be wary of the property name? For example, `cmdr` might appear in some response data already or in the future, perhaps use something more distinct like `original_request_cmdr`?
> 
> ```python
> capi_data['original_request_cmdr'] = capi_response.cmdr
> ```
> 
> Or alternatively we could pass it as a separate parameter to the plugin callback function, keeping it out of the `CAPIData`?

I prefer having it in CAPIData.  We originally made that for slightly better type enforcing, but I think it's useful to have these additional properties too.

**Comment by aussig** [comment]
Ok, will do. Shall I use the more lengthy `original_request_cmdr` or happy with `cmdr_name` perhaps?

**Comment by Athanasius** [comment]
> Ok, will do. Shall I use the more lengthy `original_request_cmdr` or happy with `cmdr_name` perhaps?

Let's split the difference with `request_cmdr`.

**Comment by aussig** [comment]
Oh FFS sorry, I had 'show whitespace' temporarily switched off in my diff viewer, so I didn't see a bunch of whitespace changes in `PLUGINS.py` that the editor applied automatically. 

I'll fix this in another commit.

**Comment by aussig** [comment]
Ok, stupid whitespace chnages reverted in documentation.

The code changes to pass `request_cmdr` in `CAPIData` are ready for review again now.

File: companion.py
Comment (Athanasius) [inline companion.py:1080]: This gets `cmdr` back into EDMarketConnector.py:AppWindow.capi_handle_response() ... but then you don't seem to have added anything to populate `capi_response.capi_data.request_cmdr` with it.

Comment (Athanasius) [reply companion.py:1080]: Overall I'd adjust `capi_single_query()` to store a copy of `monitor.cmdr` just before the actual `.get()` of the query, and then fill in `request_cmdr` *in CAPIData` when it creates that.  Do the latter by adjusting the CAPIData `__init__()` method.

i.e. it turns out there's no need to mess with EDMCAPIRequest/Response, as it goes into the CAPIData that is held in the latter.

**Review by Athanasius — CHANGES_REQUESTED** [review]

File: EDMarketConnector.py
Comment (A-UNDERSCORE-D) [inline EDMarketConnector.py:920]:
```suggestion
        new_data: dict[str, Any] = {}
```

File: EDMarketConnector.py
Comment (A-UNDERSCORE-D) [inline EDMarketConnector.py:1014]:
```suggestion
        new_data: dict[str, Any] = {}
```

File: EDMarketConnector.py
Comment (A-UNDERSCORE-D) [inline EDMarketConnector.py:1103]:
```suggestion
        new_data: dict[str, Any] = {}
```

File: EDMarketConnector.py
Comment (A-UNDERSCORE-D) [inline EDMarketConnector.py:1172]: we have deep_get for this

File: PLUGINS.md
Comment (A-UNDERSCORE-D) [inline PLUGINS.md:966]: For @Athanasius  more than anyone else

This is where I'd suggest a `match`

File: companion.py
Comment (A-UNDERSCORE-D) [inline companion.py:579]:
```suggestion
        cmdr: str | None = None
```

File: companion.py
Comment (A-UNDERSCORE-D) [inline companion.py:596]:
```suggestion
            cmdr: str | None = None
```

File: companion.py
Comment (A-UNDERSCORE-D) [inline companion.py:678]:
```suggestion
        new_data: dict[str, Any]
```

File: companion.py
Comment (A-UNDERSCORE-D) [inline companion.py:799]:
```suggestion
            new_data: dict[str, Any]
```

File: companion.py
Comment (A-UNDERSCORE-D) [inline companion.py:1055]:
```suggestion
            self, query_time: int, tk_response_event: str | None  = None,
```

File: plug.py
Comment (A-UNDERSCORE-D) [inline plug.py:406]:
```suggestion
) -> str | None:
```

File: plug.py
Comment (A-UNDERSCORE-D) [inline plug.py:420]:
```suggestion
                error = error if error else newerror
```

File: plugins/eddb.py
Comment (A-UNDERSCORE-D) [inline plugins/eddb.py:172]:
```suggestion
    new_entry: dict[str, Any]
```

File: plugins/eddn.py
Comment (A-UNDERSCORE-D) [inline plugins/eddn.py:406]:
```suggestion
        new_data: dict[str, Any]
```

File: plugins/eddn.py
Comment (A-UNDERSCORE-D) [inline plugins/eddn.py:642]:
```suggestion
        new_data: dict[str, Any]
```

There are more of these, but Im not going to cover each one :sweat_smile:

**Review by A-UNDERSCORE-D — CHANGES_REQUESTED** [review]

File: companion.py
Comment (aussig) [reply companion.py:1080]: I'm setting `capi_data['request_cmdr']` inside the `EDMCCAPIResponse` instance (`companion.py` line 600), so it's there for every instance, whatever the CAPI request that triggered it.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:920]: Might be beginning to regret incorporating this typing change into my PR ;)

File: companion.py
Comment (Athanasius) [reply companion.py:1080]: 1. That's wrong, as it shouldn't be **in the dictionary**.  It should be a property on `CAPIData` instead.
2. If we're doing it for `/fleetcarrier` there's no reason to not do it for all the CAPI queries, hence my suggestion to centralise it in `capi_single_query()`.

File: EDMarketConnector.py
Comment (Athanasius) [reply EDMarketConnector.py:920]: Yes, tbh I'm about to merge my mypy PR, and that will have fixed/changed these things anyway.  So, when you rebase on top of that, please just drop any of your commits that are only that sort of thing and not about *your* code changes.

File: plugins/eddn.py
Comment (aussig) [reply plugins/eddn.py:642]: Ok, I've changed all that I originally added in this PR, but there are many more across the codebase and I don't want to expand the footprint of my PR any further - these should be fixed in a separate PR IMHO.

File: EDMarketConnector.py
Comment (aussig) [reply EDMarketConnector.py:1172]: `deep_get` is defined in the CLI interface `EDMC.py` and is currently only used there. Should I be importing that into `EDMarketConnector.py`?

Comment (Athanasius) [reply EDMarketConnector.py:1172]: Leave it for this PR.  If we end up doing that then `deep_get()` needs to end up in `utils/deep_get.py` or somesuch, and that's not a relevant change for this work.

File: plug.py
Comment (aussig) [reply plug.py:406]: I was following the same pattern as the existing function `notify_capidata`.  Should I change the return type for that function too?  Extending that question further, the pattern `-> Optional[str]:` is used for many functions across the codebase, perhaps there should be a separate PR to review those.

File: plug.py
Comment (aussig) [reply plug.py:420]: Again following the same pattern as `notify_capidata()`.  Should that function change too?

Comment (Athanasius) [reply plug.py:420]: Yes, just ignore @A-UNDERSCORE-D for these things in this PR.  We'll do such general cleanup in other PRs.

**Comment by Athanasius** [comment]
I've been mulling over how the FC query is triggered, and just realised that triggering it *immediately* on `CarrrierStats` has the potential to delay the "Automatically update on docking" *startup* set of queries.

So, we probably want to make `CarrierStats` cause an `after` call to trigger it, rather than it being immediate.  The specific idea being to ensure it will get triggered *after* the 'station' queries on login.  We do *not* want to delay those at all.

Then there's a whole thing with possibly wanting to run any `/fleetcarrier` query in its own separate thread, so as to never get in the way of the 'station' queries, but that's infrastructure changes not appropriate for *this* PR.

**Comment by aussig** [comment]
It's already using an `after` call: `EDMarketConnector.py` line 1588:

```python
self.w.after(int(SERVER_RETRY * 1000), self.capi_request_fleetcarrier_data)
```

But do you mean somehow ensuring this is also after the 'station' queries?  How would I ensure those had happened though...

**Comment by Athanasius** [comment]
> It's already using an `after` call: `EDMarketConnector.py` line 1588:
> 
> ```python
> self.w.after(int(SERVER_RETRY * 1000), self.capi_request_fleetcarrier_data)
> ```
> 
> But do you mean somehow ensuring this is also after the 'station' queries? How would I ensure those had happened though...

Ah, I'd missed that you were using the retry timer for that, it will probably suffice.  To be after the station queries the delay just needs to be long enough.

However, another thought.  There could be FC owners with large cargo/orders data that don't run any plugin that makes use of the data.  Delaying their station queries for a long time due to a `/fleetcarrier` query taking a long time is bad UX.  So, please add an option, which will default *off*, to enable/disable the FC queries.  I would place this:

1. On the Configuration tab.
2. Above the 'Hotkey' section.

**Review by Athanasius — APPROVED** [review]

**Comment by aussig** [comment]
> However, another thought. 

You're a demanding boss ;)   Looking at the current layout of the Configuration panel, it's at the limit of available vertical space.  A couple of options here:

1. We could try to reduce vertical space used on the existing elements. A possible candidate are the horizontal separators which are currently quite tall.
2. We could introduce a scrollbar to the pane.

In both cases, I think I could make a case for this UI stuff to be separated out into a separate issue and therefore PR?

**Comment by Athanasius** [comment]
I want it in this PR so that once this is merged it's in a state where I'd be happy for it to be in a release.  Else if we merge this PR and other things mean we need to make a release we get into a messy "well, either FC owners have to put up with this, or we need to temporarily disable triggering code just for this release, but then remember to undo that after the release branch is forked off, and ensure neither accidentally merges something that changes that state, and ...".

I really don't want to make a release where (some) FC owners are suddenly in a situation where the automatic docking updates of station data just don't happen in a timely manner.  Keep in mind that all those "is it the same system/station as journal says?" checks are upon the received data being processed back in the main thread.  If the player has undocked by then due to a slow FC query then the entirety of the station queries will have been a waste.

**Comment by Athanasius** [comment]
WRT "Settings is already too tall"... you obviously don't have as many disabled plugins as I do.  Those cause the Plugins tab to be taller than any other in my setup, maybe around 55-60% of my 2560x1440 monitor height.

Please just add this option as requested and we can consider changes to accommodate smaller screens in another PR.

**Comment by aussig** [comment]
Ok, I'll get onto it...

File: EDMarketConnector.py
Comment (Athanasius) [inline EDMarketConnector.py:1585]: It'll end up *stored* as in int on Windows, but you should be using `get_bool()` and sending it a `bool` value in `config.set()`.

File: prefs.py
Comment (Athanasius) [inline prefs.py:474]: So this should be `tk.BooleanVar(...)`

File: prefs.py
Comment (Athanasius) [inline prefs.py:1266]: And you should just be able to lose the `int(...)` wrapping here.

**Review by Athanasius — CHANGES_REQUESTED** [review]

**Review by Athanasius — APPROVED** [review]

**Comment by Athanasius** [comment]
OK, I'm going to merge this and then fix up some fiddly bits, including making `Session.dump_capi_data()` happy when it's `/fleetcarrier` output.

**Pull request merged**
